runtime/near-evm-runner/src/evm_state.rs

```diff
@@ -0,0 +1,533 @@
+        self.accounts.insert(address.0, account.clone());
+        Ok(())
+    }
+
+    fn _read_contract_storage(&self, key: [u8; 52]) -> Result<Option<[u8; 32]>> {
+        let mut addr = [0u8; 20];
+        addr.copy_from_slice(&key[..20]);
+        if self.self_destructs.contains(&addr) {
+            Ok(None)
+        } else {
+            Ok(self.storages.get(&key.to_vec()).cloned())
+        }
+    }
+
+    fn _set_contract_storage(&mut self, key: [u8; 52], value: [u8; 32]) -> Result<()> {
+        self.storages.insert(key.to_vec(), value);
+        Ok(())
+    }
+
+    fn commit_changes(&mut self, other: &StateStore) -> Result<()> {
+        self.commit_self_destructs(&other.self_destructs);
+        self.commit_recreated(&other.recreated);
+        self.commit_code(&other.code);
+        self.commit_accounts(&other.accounts);
+        self.commit_storages(&other.storages);
+        self.logs.extend(other.logs.iter().cloned());
+        Ok(())
+    }
+
+    fn recreate(&mut self, addr: [u8; 20]) {
+        self.code.remove(&addr);
+        // We do not remove account because balances persist across recreation.
+        self.accounts.entry(addr).or_insert(EvmAccount::default()).nonce = U256::from(0);
+        self.overwrite_storage(addr);
+        self.self_destructs.remove(&addr);
+        self.recreated.insert(addr);
+    }
+}
+
+pub struct SubState<'a> {
+    pub msg_sender: &'a Address,
+    pub state: &'a mut StateStore,
+    pub parent: &'a dyn EvmState,
+}
+
+impl SubState<'_> {
+    pub fn new<'a>(
+        msg_sender: &'a Address,
+        state: &'a mut StateStore,
+        parent: &'a dyn EvmState,
+    ) -> SubState<'a> {
+        SubState { msg_sender, state, parent }
+    }
+
+    pub fn self_destruct(&mut self, address: &Address) -> Result<()> {
+        self.state.self_destructs.insert(address.0);
+        let mut account = self.get_account(address)?.unwrap_or_default();
+        account.nonce = U256::from(0);
+        self.state.set_account(address, &account)
+    }
+}
+
+impl EvmState for SubState<'_> {
+    fn code_at(&self, address: &Address) -> Result<Option<Vec<u8>>> {
+        let internal_addr = utils::evm_account_to_internal_address(*address);
+        if self.state.self_destructs.contains(&internal_addr) {
+            Ok(None)
+        } else {
+            self.state
+                .code
+                .get(&internal_addr)
+                .map_or_else(|| self.parent.code_at(address), |k| Ok(Some(k.to_vec())))
+        }
+    }
+
+    fn set_code(&mut self, address: &Address, bytecode: &[u8]) -> Result<()> {
+        let internal_addr = utils::evm_account_to_internal_address(*address);
+        self.state.code.insert(internal_addr, bytecode.to_vec());
+        Ok(())
+    }
+
+    fn get_account(&self, address: &Address) -> Result<Option<EvmAccount>> {
+        self.state
+            .get_account(address)?
+            .map_or_else(|| self.parent.get_account(address), |v| Ok(Some(v)))
```

Comment:
looks expensive when substate is deep

Reply:
what do you mean "is deep"?

Reply:
oh K is contract call stack that not going to be big, chain of K hash table is okay